File: examples/typescript/patterns/code-organization/separate-logic.ts
// Business logic: no DurableContext, pure work.
async function validateOrder(order: Order): Promise<ValidationResult> {
  return orderValidator.validate(order);
}

async function chargePayment(order: Order): Promise<Receipt> {
  return paymentService.charge(order.total, order.cardToken);
}

async function scheduleShipment(order: Order): Promise<ShipmentId> {
  return shipmentService.schedule(order.id, order.address);
}

// Orchestration: the handler reads as a sequence of intent.
export const handler = withDurableExecution(
  async (order: Order, context: DurableContext) => {
    await context.step("validate", () => validateOrder(order));
    const receipt = await context.step("charge", () => chargePayment(order));
    const shipmentId = await context.step("schedule", () => scheduleShipment(order));
    return { receipt, shipmentId };
  },
);

File: examples/csharp/patterns/code-organization/separate-logic.cs
using Amazon.Lambda.Core;
using Amazon.Lambda.DurableExecution;

// Business logic: no IDurableContext, pure work.
public static class OrderLogic
{
    public static ValidationResult ValidateOrder(Order order) => new(order.Id, Valid: true);

    public static Receipt ChargePayment(Order order) => new(order.Id, order.Total);

    public static string ScheduleShipment(Order order) => $"ship-{order.Id}";
}

// Orchestration: the workflow reads as a sequence of intent.
public class SeparateLogicExample
{
    public Task<DurableExecutionInvocationOutput> Handler(
        DurableExecutionInvocationInput input, ILambdaContext context)
        => DurableFunction.WrapAsync<Order, OrderResult>(Workflow, input, context);

    private async Task<OrderResult> Workflow(Order order, IDurableContext ctx)
    {
        await ctx.StepAsync(
            async (_, _) => OrderLogic.ValidateOrder(order),
            name: "validate");

        Receipt receipt = await ctx.StepAsync(
            async (_, _) => OrderLogic.ChargePayment(order),
            name: "charge");

        string shipmentId = await ctx.StepAsync(
            async (_, _) => OrderLogic.ScheduleShipment(order),
            name: "schedule");

        return new OrderResult(receipt, shipmentId);
    }
}

public record Order(string Id, decimal Total, string Address, string CardToken);
public record ValidationResult(string OrderId, bool Valid);
public record Receipt(string OrderId, decimal Amount);
public record OrderResult(Receipt Receipt, string ShipmentId);
